### PR TITLE
feature: refactor pod enhancer so all workload resource types are used

### PR DESCRIPTION
**Pull request by Jiahui-Zhang-20**
Previously:

The pod enhancer does not consider workload resource types such as `ReplicaSet` and `Deployment`. There were several  k8s API calls for retrieving throughout the project. This makes it hard to refactor and use the indexer for caching. This is because each API call has to be specific to the object type (e.g. `Pod`, `Job`, `Deployment`...)

Now:

The query logic is refactored with a function named `findObject` that generically returns `metav1.Object`. The cached object is also no longer just`interface{}` but instead `metav1.Object`.

In order to associate a pod state change to any owning objects, we use DFS to find all owning objects (usually one) of type `metav1.Object`. Then the pod enhancer calls specific owning object enhancers to add metadata to the sentry event.

### CODE REVIEW

File: enhancers_pod.go [hunk 1]
```diff
@@ -71,46 +38,225 @@ func runPodEnhancer(ctx context.Context, podMeta *v1.ObjectReference, cachedObje
...
+		case "Pod":
+			sentryEvent.Fingerprint = append(sentryEvent.Fingerprint, rootOwner.object.GetName())
+		case "ReplicaSet":
+			ReplicaSetPodEnhancer(ctx, scope, rootOwner.object, sentryEvent)
+		case "Deployment":
+			deploymentPodEnhancer(ctx, scope, rootOwner.object, sentryEvent)
+		case "Job":
+			jobPodEnhancer(ctx, scope, rootOwner.object, sentryEvent)
+		case "CronJob":
+			cronjobPodEnhancer(ctx, scope, rootOwner.object, sentryEvent)
```
Comment (tonyo) [inline enhancers_pod.go]: Do we actually need to do the switch-case here? All those `*PodEnhancer` functions look quite similar, and probably can be generalized.

From a different angle: should we instead implement standalone enhancers for Deployments, Cronjobs, etc? The idea would be:
* If we're processing an event that involves a given Deployment, we'll run the Deployment enhancer on it.
* If we're processing a pod that belongs to a given Deployment (you already have the owner detection logic in place), we'll also run the same Deployment enhancer

File: enhancers_pod.go [hunk 2]
```diff
@@ -71,46 +38,225 @@ func runPodEnhancer(ctx context.Context, podMeta *v1.ObjectReference, cachedObje
...
+	scope.AddBreadcrumb(&sentry.Breadcrumb{
+		Message:   fmt.Sprintf("Created cronjob %s", object.GetName()),
+		Level:     sentry.LevelInfo,
+		Timestamp: object.GetCreationTimestamp().Time,
+	}, breadcrumbLimit)
+
+	return nil
+}
+
+func ReplicaSetPodEnhancer(ctx context.Context, scope *sentry.Scope, object metav1.Object, sentryEvent *sentry.Event) error {
```
Comment (tonyo) [inline enhancers_pod.go]:
```suggestion
func replicaSetPodEnhancer(ctx context.Context, scope *sentry.Scope, object metav1.Object, sentryEvent *sentry.Event) error {
```
For consistency?

File: watcher_pods.go [hunk 3]
```diff
@@ -33,7 +33,7 @@ func handlePodTerminationEvent(ctx context.Context, containerStatus *v1.Containe
 	}
 
 	setTagIfNotEmpty(scope, "reason", state.Reason)
-	setTagIfNotEmpty(scope, "kind", pod.Kind)
+	setTagIfNotEmpty(scope, "kind", "Pod")
```
Comment (tonyo) [inline watcher_pods.go]: If there's no constants in client-go for resource kinds, let's declare those ourselves (e.g. will help to avoid typos).

File: enhancers.go [hunk 4]
```diff
@@ -2,29 +2,325 @@ package main
...
+
+func findRootOwners(ctx context.Context, kindObjPair *KindObjectPair) ([]KindObjectPair, error) {
+
+	parents := kindObjPair.object.GetOwnerReferences()
+	// the owners slice to be returned
+	rootOwners := []KindObjectPair{}
+
+	// base case: the object has no parents
+	if len(parents) == 0 {
+		rootOwners = append(rootOwners, *kindObjPair)
```
Comment (tonyo) [inline enhancers.go:108]: Q: why do we add the object itself if it's not managed by anything?

File: enhancers.go [hunk 5]
```diff
@@ -2,29 +2,325 @@ package main
...
+	scope.AddBreadcrumb(&sentry.Breadcrumb{
+		Message:   fmt.Sprintf("Created replicaset %s", object.GetName()),
+		Level:     sentry.LevelInfo,
+		Timestamp: object.GetCreationTimestamp().Time,
+	}, breadcrumbLimit)
+
+	return nil
+}
+
+func deploymentPodEnhancer(ctx context.Context, scope *sentry.Scope, object metav1.Object, sentryEvent *sentry.Event) error {
```
Comment (tonyo) [inline enhancers.go]: Look like *PodEnhancer functions can be refactored (do they have to be Pod-related?), but it's ok to do it in a separate PR.

**Review by tonyo — COMMENTED** [review]
A few minor comments, otherwise good stuff 👍

File: enhancers.go (hunk 4 above)
Comment (Jiahui-Zhang-20) [reply enhancers.go:108]: Essentially I'm using DFS to find the objects are have no owningReferences. If an object is not managed by anything, then it is itself a root. I think this makes the base case cleaner.

In the case the enhancer calls an object with no owning references, I just make sure I don't call the enhancer twice on the same object by checking that the object and any object in its rootOwners slice are not equal.

Comment (tonyo) [reply enhancers.go:108]: Maybe it makes sense from the DFS perspective, but interface-wise it doesn't sound clean to me. When I call the function called `findRootOwners`, I don't expect to have the object itself included in the result.

> I don't call the enhancer twice on the same object by checking that the object and any object in its rootOwners slice are not equal.

FWIW, checking by name only is not entirely safe (e.g. a pod theoretically can be owned by a resource with the same name).

File: sentry_dsn_data.go [hunk 6]
```diff
@@ -7,11 +7,21 @@ import (
...
 	"github.com/getsentry/sentry-go"
+	v1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 var DSNAnnotation = "k8s.sentry.io/dsn"
 
+const (
+	POD        string = "Pod"
```
Comment (tonyo) [inline sentry_dsn_data.go]: Here and below
```suggestion
	KindPod        string = "Pod"
```

Also let's move these to a file with a better name (e.g. `constants.go`)

File: watcher_events_test.go [hunk 7]
```diff
@@ -63,7 +63,7 @@ func TestHandleWatchEvent(t *testing.T) {
 			},
 			InvolvedObject: corev1.ObjectReference{
 				APIVersion:      "batch/v1",
-				Kind:            "CronJob",
+				Kind:            CRONJOB,
```
Comment (tonyo) [inline watcher_events_test.go]: Let's not use the constants in tests, it'll help us to catch changes in the values of those constants.

File: watcher_pods.go [hunk 8]
```diff
@@ -216,3 +214,68 @@ func startPodWatchers(ctx context.Context, config *rest.Config, namespaces []str
 
 	}
 }
+
+// Starts all informers (jobs, cronjobs, replicasets, deployments)
+// if we opt into cronjob, attach the job/cronjob event handlers
+// and add to the crons monitor data struct for Sentry Crons
+func startInformers(ctx context.Context, namespace string) error {
```
Comment (tonyo) [inline watcher_pods.go]: We should find a better place (file) for this function and `*Informer` objects. Feel free to put them in a separate file.

File: enhancers.go (hunk 4 above)
Comment (Jiahui-Zhang-20) [reply enhancers.go:108]: Makes sense. Now `findRootOwners` returns an empty slice if the object passed in has no owner references.

Comment (tonyo) [reply enhancers.go:108]: Thanks!

File: enhancers_test.go [hunk 9]
```diff
@@ -57,9 +53,9 @@ func TestRunPodEnhancer(t *testing.T) {
 	// Add event message
 	event.Message = "This event is for TestRunPodEnhancer"
 	// Call pod enhancer to modify scope and event
-	err = runPodEnhancer(ctx, objRef, nil, scope, event)
+	err = runEnhancers(ctx, nil, "Pod", podObj, scope, event)
```
Comment (tonyo) [inline enhancers_test.go]:
```suggestion
	err = runEnhancers(ctx, nil, KindPod, podObj, scope, event)
```

**Review by tonyo — APPROVED** [review]
Thanks for updating, just one non-blocking suggestion.
